app/models/concerns/user_omniauth_concern.rb

```diff
@@ -0,0 +1,34 @@
+module UserOmniauthConcern
+  extend ActiveSupport::Concern
+  include UserOmniauthFacebookConcern
+
+  included do
+    devise :omniauthable
+  end
+
+  module ClassMethods
+    # Find or create a user from omniauth data.
+    #
+    # @param [OmniAuth::AuthHash] auth The data with omniauth info.
+    # @return [User] The user found or created from omniauth data.
+    def from_omniauth(auth)
+      identity = User::Identity.find_by(provider: auth.provider, uid: auth.uid)
+      return identity.user if identity
+
+      User.create do |user|
+        user.assign_attributes(name: auth.info.name, email: auth.info.email,
+                               password: Devise.friendly_token[0, 20])
+        user.identities.build(provider: auth.provider, uid: auth.uid)
+        user.instance_users.build
+      end
+    end
+
+    # A convenience method that receives both parameters and session to initialize a user.
+    # This is a override of `Devise::Models::Registerable::ClassMethods#new_with_session`
+    def new_with_session(params, session)
+      super.tap do |user|
+        user.instance_users.build
```

Comment:
`from_omniauth` directly created the user and skipped the devise registrable controller actions, so I need to especially build an instance_user for it.